L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -1,52 +1,100 @@
+        bool isSeed[triggerCellsHandle->size()];
+
+        /* search for cluster seeds */
+        int itc=0;
+        for( l1t::HGCalTriggerCellBxCollection::const_iterator tc = triggerCellsHandle->begin(); tc != triggerCellsHandle->end(); ++tc,++itc ){
+            isSeed[itc] = ( tc->hwPt() > seedThreshold_) ? true : false;
+        }
+    
+        /* clustering the TCs */
+        std::vector<l1t::HGCalCluster> clustersTmp;
+
+        itc=0;
+        for( l1t::HGCalTriggerCellBxCollection::const_iterator tc = triggerCellsHandle->begin(); tc != triggerCellsHandle->end(); ++tc,++itc ){
+            
+            if( tc->hwPt() < triggerCellThreshold_ ){
+                continue;
+            }
+        
+            /* searching for TC near the center of the cluster  */
+            int iclu=0;
+            vector<int> tcPertinentClusters; 
+            for( std::vector<l1t::HGCalCluster>::iterator clu = clustersTmp.begin(); clu != clustersTmp.end(); ++clu,++iclu ){
+                if( this->isPertinent(*tc, *clu, dr_) ){
+                    tcPertinentClusters.push_back(iclu);
                 }
-                               
-                C2d_pt += tc->pt();                        
-                protoClEta += tc->pt()*tc->eta();
-                protoClPhi += tc->pt()*tc->phi();
             }
+            if( tcPertinentClusters.size() == 0 && isSeed[itc] ){
+                l1t::HGCalCluster obj( *tc );
+                clustersTmp.emplace_back( obj );
+            }
+            else if ( tcPertinentClusters.size() > 0 ){
+         
+                unsigned minDist = 300;
+                unsigned targetClu = 0;
+                        
+                for( std::vector<int>::const_iterator iclu = tcPertinentClusters.begin(); iclu != tcPertinentClusters.end(); ++iclu ){
+                    double d = clustersTmp.at( *iclu).distance(*tc);
+                    if( d < minDist ){
+                        minDist = d;
+                        targetClu = *iclu;
+                    }
+                } 
+                clustersTmp.at( targetClu).addTriggerCell( *tc );
+                edm::Ptr<l1t::HGCalTriggerCell> p( triggerCellsHandle, itc, true );
+                clustersTmp.at( targetClu).addTriggerCellList( p );
+            
+            }
+
         }
-        l1t::HGCalCluster cluster( reco::LeafCandidate::LorentzVector(), C2d_hwPtEm + C2d_hwPtHad, 0, 0);
-        cluster.setHwPtEm(C2d_hwPtEm);
-        cluster.setHwPtHad(C2d_hwPtHad);
-
-        if((cluster.hwPtEm()+cluster.hwPtHad())>tc_CUT_){
-            C2d_eta = protoClEta/C2d_pt;
-            C2d_phi = protoClPhi/C2d_pt;                
-            math::PtEtaPhiMLorentzVector calibP4(C2d_pt, C2d_eta, C2d_phi, 0 );
-            cluster.setP4(calibP4);
-            cluster_product_.push_back(0,cluster);
+
+        /* making the collection of clusters */
+        for( unsigned i(0); i<clustersTmp.size(); ++i ){
+            if( clustersTmp.size()>0 ){
```

Comment:
In any case, this is not needed. If you enter the loop it means that the size is >0